service/worker/deployment/deployment_reachability.go

```diff
@@ -0,0 +1,183 @@
+	"go.temporal.io/server/common/worker_versioning"
+)
+
+const (
+	reachabilityCacheOpenWFsTTL   = time.Minute
+	reachabilityCacheClosedWFsTTL = 10 * time.Minute
+	reachabilityCacheMaxSize      = 10000
+)
+
+// We update the reachability search attribute on workflow task completion and when there is
+// an UpdateWorkflowExecutionOptions request that changes the workflow's effective behavior
+// or deployment. We do not update the reachability search attribute if the search attribute
+// has not changed. We also do not update the reachability search attribute if a workflow has
+// started transitioning to a new deployment, because only on task completion do we know whether
+// the transition has succeeded or failed.
+func getDeploymentReachability(
+	ctx context.Context,
+	nsId, nsName, seriesName, buildID, currentBuildID string,
+	currentBuildIDValidTime time.Time,
+	rc reachabilityCache,
+) (enumspb.DeploymentReachability, time.Time, error) {
+	// 1a. Reachable by new unpinned workflows
+	if buildID == currentBuildID { // add if buildID is ramping, once we have ramp
+		return enumspb.DEPLOYMENT_REACHABILITY_REACHABLE, currentBuildIDValidTime, nil
+	}
+
+	// 2a. Reachable by open pinned workflows
+	countRequest := makeCountRequest(nsId, nsName, seriesName, buildID, true)
+	exists, lastUpdateTime, err := rc.Get(ctx, countRequest, true)
+	if err != nil {
+		return enumspb.DEPLOYMENT_REACHABILITY_UNSPECIFIED, time.Time{}, err
+	}
+	if exists {
+		return enumspb.DEPLOYMENT_REACHABILITY_REACHABLE, lastUpdateTime, nil
+	}
+
+	// 3. Reachable by closed pinned workflows
+	countRequest = makeCountRequest(nsId, nsName, seriesName, buildID, false)
+	exists, lastUpdateTime, err = rc.Get(ctx, countRequest, false)
+	if err != nil {
+		return enumspb.DEPLOYMENT_REACHABILITY_UNSPECIFIED, time.Time{}, err
+	}
+	if exists {
+		return enumspb.DEPLOYMENT_REACHABILITY_CLOSED_WORKFLOWS_ONLY, lastUpdateTime, nil
+	}
+
+	// 4. Unreachable
+	return enumspb.DEPLOYMENT_REACHABILITY_UNREACHABLE, time.Now(), nil
+}
+
+func makeCountRequest(
+	namespaceId, namespaceName, seriesName, buildId string, open bool,
+) manager.CountWorkflowExecutionsRequest {
+	return manager.CountWorkflowExecutionsRequest{
+		NamespaceID: namespace.ID(namespaceId),
+		Namespace:   namespace.Name(namespaceName),
+		Query:       makeDeploymentQuery(seriesName, buildId, open),
+	}
+}
+
+func makeDeploymentQuery(seriesName, buildID string, open bool) string {
+	var statusFilter string
+	deploymentFilter := "= " + worker_versioning.ReachabilityBuildIdSearchAttribute(enumspb.VERSIONING_BEHAVIOR_PINNED, &deploymentpb.Deployment{
+		SeriesName: seriesName,
+		BuildId:    buildID,
+	})
+	if open {
+		statusFilter = "= 'Running'"
+	} else {
+		statusFilter = "!= 'Running'"
+	}
+	// todo (carly) part 2: handle null search attribute / unversioned deployment
+	return fmt.Sprintf("%s %s AND %s %s", searchattribute.BuildIds, deploymentFilter, searchattribute.ExecutionStatus, statusFilter)
```

Comment:
what happens if there is space or quotes in the series name or build ID? do you need to use something like `sqlparser.String(sqlparser.NewStrVal([]byte(...)))`

Reply:
I used `sqlparser.String(sqlparser.NewStrVal([]byte(...)))` on the whole `deploymentFilter` I think.. will double check, but I think that should handle it

Reply:
oh you did, sorry I missed that.